internal/output/output_result.go

```diff
@@ -0,0 +1,538 @@
+type OutputPackageResult struct { //nolint:revive
+	Name             string
+	InstalledVersion string
+	FixedVersion     string
+	CalledVulns      []OutputVulnResult
+	UncalledVulns    []OutputVulnResult
+	LayerDetail      OutputPackageLayerDetail
+	VulnCount        OutputVulnCount
+}
+
+// OutputVulnResult represents a single vulnerability.
+type OutputVulnResult struct { //nolint:revive
+	ID             string
+	GroupIDs       []string
+	Aliases        []string
+	FixedVersion   string
+	SeverityRating string
+	SeverityScore  string
+}
+
+// OutputPackageLayerDetail represents detailed layer tracing information about a package.
+type OutputPackageLayerDetail struct { //nolint:revive
+	LayerCommand        string
+	LayerCommandTooltip string
+	LayerID             string
+	InBaseImage         bool
+}
+
+type OutputLayerInfo struct { //nolint:revive
+	Index        int
+	LayerCommand string
+	LayerID      string
+	Count        OutputVulnCount
+}
+
+// OutputVulnCount represents the counts of vulnerabilities by severity and fixed/unfixed status
+type OutputVulnCount struct { //nolint:revive
+	SeverityCount     OutputSeverityCount
+	CallAnalysisCount OutputCallAnalysisCount
+	FixableCount      OutputFixableCount
+}
+
+type OutputSeverityCount struct { //nolint:revive
+	Critical int
+	High     int
+	Medium   int
+	Low      int
+	Unknown  int
+}
+
+type OutputVulnTypeCount struct { //nolint:revive
+	All      int
+	OS       int
+	Project  int
+	Uncalled int
+}
+
+type OutputCallAnalysisCount struct { //nolint:revive
+	Called   int
+	Uncalled int
+}
+
+type OutputFixableCount struct { //nolint:revive
+	Fixed   int
+	UnFixed int
+}
+
+// PrintOutputResults prints the output to the outputWriter.
+// This function is for testing purposes only, to visualize the result format.
+func PrintOutputResults(vulnResult *models.VulnerabilityResults, outputWriter io.Writer) error {
+	encoder := json.NewEncoder(outputWriter)
+	encoder.SetIndent("", "  ")
+	result := BuildOutputResults(vulnResult)
+	//nolint:musttag
+	return encoder.Encode(result)
+}
+
+// BuildOutputResults constructs the output result structure from the vulnerability results.
+//
+// This function creates a hierarchical representation of the results, starting from the overall
+// summary and drilling down to ecosystems, sources, packages, and vulnerability details.
+// This structured format facilitates generating various output formats (e.g., table, HTML, etc.).
+func BuildOutputResults(vulnResult *models.VulnerabilityResults) OutputResult {
+	var ecosystemMap = make(map[string][]OutputSourceResult)
+	var resultCount OutputVulnCount
+
+	for _, packageSource := range vulnResult.Results {
+		// Temporary workaround: it is a heuristic to ignore installed packages
+		// which are already covered by OS-specific vulnerabilities.
+		// This filtering should be handled by the container scanning process.
+		// TODO(gongh@): Revisit this after container scanning supports comprehensive functionality.
```

Comment:
nit: let's be clearer here. "Revisit this once container scanning can distinguish these cases".